File: S/839-similar-string-groups/main.go
package main

import (
	"fmt"
	"sort"
)

func numSimilarGroups(strs []string) int {
	M := map[int][]int{}
	for i := 0; i < len(strs); i++ {
		target := strs[i]
		for j := i + 1; j < len(strs); j++ {
			candi := strs[j]
			isFirst := true
			isSecond := true
			isTarget := false
			diff1, diff2 := byte('0'), byte('0')
			for n := 0; n < len(target); n++ {
				if target[n] != candi[n] {
					if isFirst {
						isFirst = false
						diff1 = target[n]
						diff2 = candi[n]
					} else if isSecond && diff1 == candi[n] && diff2 == target[n] {
						isSecond = false
						isTarget = true
					} else {
						isTarget = false
						break
					}
				}
			}
			if isTarget || isFirst {
				_, ok1 := M[i]
				if !ok1 {
					M[i] = []int{}
				}
				M[i] = append(M[i], j)
				_, ok2 := M[j]
				if !ok2 {
					M[j] = []int{}
				}
				M[j] = append(M[j], i)
			}
		}
	}
	var dfs func(buf []int, M map[int][]int, index, parent int)
	dfs = func(buf []int, M map[int][]int, index, parent int) {
		for i := 0; i < len(M[index]); i++ {
			if buf[M[index][i]] != -1 {
				continue
			}
			buf[M[index][i]] = parent
			dfs(buf, M, M[index][i], parent)
		}
	}
	buf := make([]int, len(strs))
	for i := 0; i < len(strs); i++ {
		buf[i] = -1
	}
	for i := 0; i < len(strs); i++ {
		if buf[i] != -1 {
			continue
		}
		buf[i] = i
		dfs(buf, M, i, i)
	}
	sort.Ints(buf)
	cnt := 1
	base := buf[0]
	for i := 1; i < len(buf); i++ {
		if buf[i] == base {
			continue
		}
		cnt++
		base = buf[i]
	}
	return cnt
}

func main() {
	fmt.Println()
}

File: S/839-similar-string-groups/main.rs
struct Solution {}

use std::collections::HashMap;
impl Solution {
  pub fn dfs(buf: &mut Vec<i32>, m: &HashMap<usize, Vec<usize>>, index: usize, parent: usize) {
    if !m.contains_key(&index) {
      return;
    }
    let children = m.get(&index).unwrap();
    for i in 0..children.len() {
      if buf[children[i]] != -1 {
        continue;
      }
      buf[children[i]] = parent as i32;
      Solution::dfs(buf, m, children[i], parent);
    }
  }

  pub fn num_similar_groups(strs: Vec<String>) -> i32 {
    let mut m: HashMap<usize, Vec<usize>> = HashMap::new();
    for i in 0..strs.len() {
      let target = strs[i].as_bytes();
      for j in i + 1..strs.len() {
        let candi = strs[j].as_bytes();
        let mut is_first = true;
        let mut is_second = true;
        let mut is_target = false;
        let mut diff1: u8 = '0' as u8;
        let mut diff2: u8 = '0' as u8;

        for n in 0..target.len() {
          if target[n] != candi[n] {
            if is_first {
              is_first = false;
              diff1 = target[n];
              diff2 = candi[n];
            } else if is_second && diff1 == candi[n] && diff2 == target[n] {
              is_second = false;
              is_target = true;
            } else {
              is_target = false;
              break;
            }
          }
        }
        if is_target || is_first {
          if !m.contains_key(&i) {
            m.insert(i, vec![j]);
          } else {
            m.get_mut(&i).unwrap().push(j);
          }
          if !m.contains_key(&j) {
            m.insert(j, vec![i]);
          } else {
            m.get_mut(&j).unwrap().push(i);
          }
        }
      }
    }

    let mut buf: Vec<i32> = vec![-1; strs.len()];
    for i in 0..strs.len() {
      if buf[i] != -1 {
        continue;
      }
      buf[i] = i as i32;
      Solution::dfs(&mut buf, &m, i, i);
    }

    let mut r: HashMap<usize, usize> = HashMap::new();
    for i in 0..buf.len() {
      if !r.contains_key(&(buf[i] as usize)) {
        r.insert(buf[i] as usize, i);
      }
    }
    r.len() as i32
  }
}

fn main() {
  println!(
    "{}",
    Solution::num_similar_groups(vec![
      String::from("tars"),
      String::from("rats"),
      String::from("arts"),
      String::from("star")
    ])
  );
}
